plugins/backend-trace-forwarder/src/Cardano/BM/Backend/TraceForwarder.lhs

```diff
@@ -248,48 +262,50 @@ spawnDispatcher config tfMVar = do
+        -- so there's no need to read items from the queue, just try to establish
+        -- the new connection.
+        void $ Async.async $ establishConnection tfTVar
+      Just h -> do
+        -- The handle is here: the connection is established and probably alive,
+        -- so read items from the queue and try to write them in the handle.
+        itemsList <- atomically $ TBQ.flushTBQueue (tfQueue currentTF)
+        sendItems config tfTVar h itemsList
 
 -- Try to send log items to the handle.
 sendItems :: ToJSON a
           => Configuration
-          -> TraceForwarderMVar a
+          -> TraceForwarderTVar a
+          -> Handle
           -> [LogObject a]
-          -> IO [LogObject a]
           -> IO ()
-sendItems _ _ [] _ = return ()
-sendItems config tfMVar items@(lo:_) getStateDigest =
-  tfHandle <$> readMVar tfMVar >>= \case
-    Nothing -> do
-      -- There's no handle, initiate the connection.
-      establishConnection 1 1 tfMVar
-      -- Connection is re\-established, try to send log item.
-      -- Since the connection is re\-established, add state digest items as well.
-      additionalItems <- getStateDigest
-      let allItems = additionalItems ++ items
-      sendItems config tfMVar allItems getStateDigest
-    Just h ->
-      try (BSC.hPutStrLn h $! encodedHostname) >>= \case
+sendItems _ _ _ [] = return ()
+sendItems config tfTVar h items@(lo:_) =
+  try (BSC.hPutStrLn h $! encodedHostname) >>= \case
+    Right _ ->
+      -- Hostname was written to the handler successfully,
+      -- try to write serialized list of LogObjects.
+      try (BSC.hPutStrLn h $! bs) >>= \case
         Right _ ->
-          -- Hostname was written to the handler successfully,
-          -- try to write serialized list of LogObjects.
-          try (BSC.hPutStrLn h $! bs) >>= \case
-            Right _ ->
-              return () -- Everything is ok, LogObjects were written to the handler.
-            Left (_e :: IOException) -> do
-              reConnectIfQueueIsAlmostFull
-              threadDelay 10000
-              sendItems config tfMVar items getStateDigest
+          return () -- Everything is ok, LogObjects were written to the handler.
         Left (_e :: IOException) -> do
-          reConnectIfQueueIsAlmostFull
-          threadDelay 10000
-          sendItems config tfMVar items getStateDigest
+          -- Handle is bad, it looks like the connection is already broken.
+          -- Remove bad handle to initiate reconnection.
+          -- Old handle will be removed later by GC.
+          atomically $ modifyTVar' tfTVar $ \be -> be { tfHandle = Nothing }
+          -- Spawn new thread to establish new connection.
+          void $ Async.async $ establishConnection tfTVar
+    Left (_e :: IOException) -> do
+      -- Handle is bad, it looks like the connection is already broken.
+      -- Remove bad handle to initiate reconnection.
+      -- Old handle will be removed later by GC.
```

Comment:
We should reference the documentation here: https://hackage.haskell.org/package/base-4.14.0.0/docs/System-IO.html#t:Handle

Reply:
Perhaps we could add a note on GC handling of Handles in this file, and reference this local note.

See section "Comments and notes" in chapter 5.6 "Developing GHC" of the AOSA book: https://aosabook.org/en/ghc.html

Reply:
Done.